text/0035-schema-applies-to-unspecified.md

```diff
@@ -0,0 +1,167 @@
+# Change the Semantics `appliesTo` Fields in the Schema
+
+## Related issues and PRs
+
+- Reference Issues: [cedar#351](https://github.com/cedar-policy/cedar/issues/351)
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2023-10-25
+
+## Summary
+
+| Scenario | Status Quo | Proposal |
+| --------- | -------- | ------- |
+| Specified `appliesTo` but omitted `principalTypes` and/or `resourceTypes` | Authz request with this action can only have an unspecified entity as the principal and/or resource | Same as status quo |
+| `appliesTo.principalTypes: []` AND `appliesTo.resourceTypes: []` | Action cannot be used in authz request | Same as status quo |
+| `appliesTo.principalTypes: []` OR `appliesTo.resourceTypes: []` | Action cannot be used in authz request | ❌ Not allowed |
```

Comment:
Based on the comment that we're concerned about

> users that currently have appliesTo.principalTypes: [] and/or appliesTo.resourceTypes: [] in their schemas

This proposal is backwards-compatible for users that have BOTH `principalTypes` and `resourceTypes` set to `[]`, but not for users that have only ONE of those set to `[]`, for a given action.

Are we concerned about that backwards-compatibility problem?  What makes us confident that we need to address backwards-compatibility for the AND case, but not for the OR case?

Reply:
In the case of `appliesTo.principalTypes: []` OR `appliesTo.resourceTypes: []`, users can proactively switch to `appliesTo.principalTypes: []` AND `appliesTo.resourceTypes: []`, and know that their schema will continue to behave as expected.

In Alternative A, the sharp edge is that in some cases there is no way for a user to proactively update their schema, because there is no way to write a schema with the same behavior before & after the change.

Reply:
I see, makes sense.  It's not backwards-compatible in the OR case, but at least there is a reasonable migration story.

(Maybe we should make up a term for the kind of backwards-incompatible that's in Alternative A, it's like a strictly worse kind of backwards-incompatible than the kind here for the OR case.)